### PR TITLE
[Vertical compaction 4/5] support partial columns write in beta rowset writer

### PR DESCRIPTION
**Pull request by wyb**
#1112 

Refact BetaRowsetWriter and add a new VerticalBetaRowsetWriter to support partical columns write.

### CODE REVIEW

**Comment by sduzh** [comment]
Hi @wyb, great job! I don't have many options for the general design, just wonder if we can keep the `BetaRowsetWriter` and `SegmentWriter` unchanged, and design a new `VerticalBetaRowsetWriter` and `VerticalSegmentWriter`,  I think this will give us more clear interfaces.
It seems that `BetaRowsetWriter` and `SegmentWriter` contain many logics we don't need here, if I'm right, it won't cause you too much work to design a new `VerticalBetaRowsetWriter` and `VerticalSegmentWriter`, correct me if I'm wrong.

File: be/src/storage/vectorized/compaction.cpp [hunk 1]
```diff
@@ -216,6 +288,80 @@ Status Compaction::merge_rowsets(int64_t mem_limit, Statistics* stats_output) {
...
+                } else {
+                    return Status::InternalError("reader get_next error.");
+                }
+            }
+
+            ChunkHelper::padding_char_columns(char_field_indexes, schema, _tablet->tablet_schema(), chunk.get());
+
+            OLAPStatus olap_status = _output_rs_writer->add_chunk_by_columns(*chunk, _column_groups[i], is_key);
+            if (olap_status != OLAP_SUCCESS) {
+                LOG(WARNING) << "writer add chunk by columns error, err=" << olap_status;
```
Comment (DeepThinker666) [inline be/src/storage/vectorized/compaction.cpp:329]: better to add tablet id or some extra info to log

File: be/src/storage/rowset/segment_v2/segment_writer.cpp [hunk 2]
```diff
@@ -263,13 +316,43 @@ Status SegmentWriter::append_chunk(const vectorized::Chunk& chunk) {
...
+    return Status::OK();
+}
+
+Status SegmentWriter::append_chunk(const vectorized::Chunk& chunk, const std::vector<uint32_t>& column_indexes,
+                                   bool is_key) {
+    DCHECK_EQ(chunk.num_columns(), column_indexes.size());
+    for (size_t i = 0; i < column_indexes.size(); ++i) {
+        uint32_t column_index = column_indexes[i];
+        if (column_index >= _column_writers.size()) {
+            return Status::InternalError("column index error");
```
Comment (DeepThinker666) [inline be/src/storage/rowset/segment_v2/segment_writer.cpp:336]: add column_index and column_writers.size() to msg

File: be/src/storage/rowset/segment_v2/segment_writer.cpp [hunk 3]
```diff
@@ -263,13 +316,43 @@ Status SegmentWriter::append_chunk(const vectorized::Chunk& chunk) {
...
+    if (is_key) {
+        for (size_t i = 0; i < chunk_num_rows; i++) {
+            // At the begin of one block, so add a short key index entry
+            if ((_num_rows_written % _opts.num_rows_per_block) == 0) {
+                size_t keys = _tablet_schema->num_short_key_columns();
+                vectorized::SeekTuple tuple(*chunk.schema(), chunk.get(i).datums());
+                std::string encoded_key = tuple.short_key_encode(keys, 0);
+                RETURN_IF_ERROR(_index_builder->add_item(encoded_key));
+            }
+            ++_num_rows_written;
```
Comment (DeepThinker666) [inline be/src/storage/rowset/segment_v2/segment_writer.cpp:352]: the increment of _num_rows_written can be done outside the is_key branch, with else together

File: be/src/storage/rowset/beta_rowset_writer.cpp [hunk 4]
```diff
@@ -221,6 +221,44 @@ OLAPStatus BetaRowsetWriter::flush() {
...
+    for (auto& segment_writer : _segment_writers) {
+        uint64_t segment_size = 0;
+        Status s = segment_writer->finalize_footer(&segment_size);
+        if (!s.ok()) {
+            LOG(WARNING) << "Fail to finalize segment footer, " << s.to_string();
+            return OLAP_ERR_WRITER_DATA_WRITE_ERROR;
+        }
+        {
+            std::lock_guard<std::mutex> l(_lock);
+            _total_data_size += segment_size;
```
Comment (DeepThinker666) [inline be/src/storage/rowset/beta_rowset_writer.cpp:241]: is it necessary to use lock?  is atomic variable enough?

**Comment by wyb** [comment]
> Hi @wyb, great job! I don't have many options for the general design, just wonder if we can keep the `BetaRowsetWriter` and `SegmentWriter` unchanged, and design a new `VerticalBetaRowsetWriter` and `VerticalSegmentWriter`, I think this will give us more clear interfaces. It seems that `BetaRowsetWriter` and `SegmentWriter` contain many logics we don't need here, if I'm right, it won't cause you too much work to design a new `VerticalBetaRowsetWriter` and `VerticalSegmentWriter`, correct me if I'm wrong.

Ok, i refact BetaRowsetWriter and add a new VerticalBetaRowsetWriter to support partical columns write.

File: be/src/storage/rowset/segment_v2/segment_writer.cpp (hunk 3 above)
Comment (wyb) [reply be/src/storage/rowset/segment_v2/segment_writer.cpp]: Write short key requires `++ _num_rows_written`

**Comment by chaoyli** [comment]
It's better to give different commit message about different pull request.

File: be/src/storage/rowset/beta_rowset_writer.cpp [hunk 5]
```diff
@@ -505,107 +616,204 @@ Status BetaRowsetWriter::_flush_src_rssids() {
...
+            // message when we encounter an error.
+            auto st = _context.env->delete_file(path);
+            LOG_IF(WARNING, !st.ok()) << "Fail to delete file=" << path << ", " << st.to_string();
+        }
     }
-    if (_segment_writer == nullptr) {
-        return OLAP_ERR_INIT_FAILED;
+}
+
+OLAPStatus VerticalBetaRowsetWriter::add_chunk_by_columns(const vectorized::Chunk& chunk,
```
Comment (sduzh) [inline be/src/storage/rowset/beta_rowset_writer.cpp:638]: IHMO, the name `add_chunk_by_columns` is somehow confusing, how about `add_columns` or `add_selected_columns` or something better?

Comment (wyb) [reply be/src/storage/rowset/beta_rowset_writer.cpp]: ok

**Review by chaoyli — APPROVED** [review]

File: be/src/storage/rowset/beta_rowset_writer.cpp [hunk 6]
```diff
@@ -53,13 +53,121 @@ BetaRowsetWriter::BetaRowsetWriter(const RowsetWriterContext& context)
...
+        if (_segment_has_deletes.size() >= 1 && _segment_has_deletes[0]) {
+            _rowset_meta->set_num_delete_files(1);
+        } else {
+            _rowset_meta->set_num_delete_files(0);
+        }
+        _rowset_meta->set_segments_overlap(NONOVERLAPPING);
+    } else {
+        if (_num_segment <= 1) {
+            _rowset_meta->set_segments_overlap(NONOVERLAPPING);
+        }
```
Comment (sduzh) [inline be/src/storage/rowset/beta_rowset_writer.cpp:123]: else?

Comment (wyb) [reply be/src/storage/rowset/beta_rowset_writer.cpp:123]: Segments overlap is set according to the context when writer inits.
This is the special case for non overlapping.

**Review by sduzh — APPROVED** [review]

**Pull request merged**